fluidkeys/keyrotate.go

```diff
@@ -152,17 +160,42 @@ func (p *alwaysFailPasswordPrompter) promptForPassword(key *pgpkey.PgpKey) (stri
 	return "", fmt.Errorf("can't prompt for password when running unattended")
 }
 
-func runKeyRotate(keys []pgpkey.PgpKey, prompter promptYesNoInterface, passwordPrompter promptForPasswordInterface) exitCode {
+type outputterInterface interface {
+	print(message string)
+	printBuffer()
+}
+
+type terminalOutputter struct{}
+
+func (o *terminalOutputter) print(message string) {
+	fmt.Print(message)
+}
+
+func (o *terminalOutputter) printBuffer() {}
+
+type cronOutputter struct {
+	buffer string
+}
+
+func (o *cronOutputter) print(message string) {
+	o.buffer += message
+}
+
+func (o *cronOutputter) printBuffer() {
```

Comment:
I'm not sure what `printBuffer` means... could be:
* print *into* buffer (not to console)
* print *the* buffer (dump whatever's in the buffer to console)

Reply:
Yep, good shout. Renaming to `printTheBuffer`

Reply:
Yep, good shout. Renaming to `printTheBuffer`

> Been thinking about this on and off over the weekend, got some random thoughts:
> 
> * any printing that happens _before_ we start using the cron outputter (e.g. things that happen in the `init` functions) won't be intercepted. I think we've got 2 options here:
>   
>   1. don't do any output in `init` functions (except for if we're about to panic)
>   2. move the init functions into `main`, right after parsing `--cron-output`
> * maybe this is the time to lay the groundwork for sending _all_ output through our own outputter object? e..g. have a _global_ formatter object (interface) which is set up in `init`.
>   
>   * The `--cron-output` flag could switch out the default outputter with the cron outputter.
>   * Key rotate can just use `Out.Print(...)` on the global pointer rather than passing an interface around

Great suggestion. Have implemented. It's nice only having fmt.Print() in one place now!

>   * Q: do packages have access to a global defined in the `main` package?

Not as far as I'm aware... Therefore I've had a go at creating a new package `out` that holds the `Outputter` and has two functions `Print` and `PrintTheBuffer`. I'm not sure this is quite right, but it's working.

Reply:
Yep, good shout. Renaming to `printTheBuffer`

> Been thinking about this on and off over the weekend, got some random thoughts:
> 
> * any printing that happens _before_ we start using the cron outputter (e.g. things that happen in the `init` functions) won't be intercepted. I think we've got 2 options here:
>   
>   1. don't do any output in `init` functions (except for if we're about to panic)
>   2. move the init functions into `main`, right after parsing `--cron-output`
> * maybe this is the time to lay the groundwork for sending _all_ output through our own outputter object? e..g. have a _global_ formatter object (interface) which is set up in `init`.
>   
>   * The `--cron-output` flag could switch out the default outputter with the cron outputter.
>   * Key rotate can just use `Out.Print(...)` on the global pointer rather than passing an interface around

Great suggestion. Have implemented. It's nice only having fmt.Print() in one place now!

>   * Q: do packages have access to a global defined in the `main` package?

Not as far as I'm aware... Therefore I've had a go at creating a new package `out` that holds the `Outputter` and has two functions `Print` and `PrintTheBuffer`. I'm not sure this is quite right, but it's working.